content/blog/aws-serverless-analytics/index.md

```diff
@@ -0,0 +1,604 @@
+
+```typescript
+// Configure batch input table 'aggregates'
+const aggregateTableName = "aggregates";
+
+const aggregateTableColumns = [
+    {
+        name: "event_type",
+        type: "string",
+    },
+    {
+        name: "count",
+        type: "int",
+    },
+    {
+        name: "time",
+        type: "string",
+    },
+];
+
+// function that lambda runs.
+// writes records to table that summarize clicks and impressions
+const aggregationFunction = async (event: EventRuleEvent) => {
+    // configure an athena client for our DW
+    const athena = require("athena-client");
+    const bucketUri = `s3://${athenaResultsBucket.get()}`;
+    const clientConfig = {
+        bucketUri,
+    };
+    const awsConfig = {
+        region,
+    };
+    const athenaClient = athena.createClient(clientConfig, awsConfig);
+    const date = moment(event.time);
+    // query impressions and clicks tables for hourly event counts
+    const partitionKey = date.utc().format("YYYY/MM/DD/HH");
+    const getAggregateQuery = (table: string) =>
+        `select count(*) from ${databaseName.get()}.${table} ` +
+        `where inserted_at='${partitionKey}'`;
+    const clicksPromise =
+        athenaClient.execute(getAggregateQuery(clicksTableName))
+        .toPromise();
+    const impressionsPromise =
+        athenaClient.execute(getAggregateQuery(impressionsTableName))
+        .toPromise();
+
+    const clickRows = await clicksPromise;
+    const impressionRows = await impressionsPromise;
+    // write the event counts to our table in S3 in JSON format
+    const clickCount = clickRows.records[0]["_col0"];
+    const impressionsCount = impressionRows.records[0]["_col0"];
+    const data = `{ "event_type": "${clicksTableName}", `+
+        `"count": ${clickCount}, "time": "${partitionKey}" }\n` +
+        `{ "event_type": "${impressionsTableName}", ` +
+        `"count": ${impressionsCount}, "time": "${partitionKey}"}`;
+    const s3Client = new S3();
+    await s3Client.putObject({
+        Bucket: dwBucket.get(),
+        Key: `${aggregateTableName}/${partitionKey}/results.json`,
+        Body: data,
+    }).promise();
+};
+
+//configure resource access policies required by the function
+const policyARNsToAttach: pulumi.Input<ARN>[] = [
+    aws.iam.ManagedPolicies.AmazonAthenaFullAccess,
+    aws.iam.ManagedPolicies.AmazonS3FullAccess,
+];
+
+const aggregateTableArgs: BatchInputTableArgs = {
+    columns: aggregateTableColumns,
+    jobFn: aggregationFunction,
+    scheduleExpression,
+    policyARNsToAttach,
+    dataFormat: "JSON",
+};
+
+dataWarehouse.withBatchInputTable(aggregateTableName, aggregateTableArgs);
+```
+
+![Athena Aggregate Table Query Results](./AggregatesExample.png)
+
+While Lambda’s limits restrict the utility of this method, the concept is powerful. Pulumi gives you the building blocks to define your abstractions. Rather than Lambda, the execution plane could be ECS, Kubernetes, or EMR.
```

Comment:
```suggestion
While Lambda’s limits restrict the utility of this method, the concept is powerful. Pulumi gives you the building blocks to define your abstractions. Instead of Lambda, the execution plane could be ECS, Kubernetes, or EMR.
```

Reply:
👍